dispatch/dispatch_test.go

```diff
@@ -684,6 +701,80 @@ func TestDispatcherRaceOnFirstAlertNotDeliveredWhenGroupWaitIsZero(t *testing.T)
 	require.Len(t, recorder.Alerts(), numAlerts)
 }
 
+func TestDispatcherReceiveAndNotifyRepeatedResolvedAlerts(t *testing.T) {
+	// More information here: https://github.com/prometheus/alertmanager/pull/3006
+
+	logger := log.NewNopLogger()
+	// logger := log.NewLogfmtLogger(os.Stdout)
+	marker := types.NewMarker(prometheus.NewRegistry())
+	alerts, err := mem.NewAlerts(context.Background(), marker, time.Hour, nil, logger, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer alerts.Close()
+
+	route := &Route{
+		RouteOpts: RouteOpts{
+			Receiver:       "default",
+			GroupBy:        map[model.LabelName]struct{}{model.AlertNameLabel: {}},
+			GroupWait:      0,
+			GroupInterval:  6 * time.Second,
+			RepeatInterval: 6 * time.Second,
+		},
+	}
+
+	recorder := &recordStage{alerts: make(map[string]map[model.Fingerprint]*types.Alert)}
+	notifier := newNotifyStage()
+	dispatcher := NewDispatcher(alerts, route, notify.FanoutStage{recorder, notifier}, marker, nil, nil, logger, NewDispatcherMetrics(false, prometheus.NewRegistry()))
+	go dispatcher.Run()
+	defer dispatcher.Stop()
+
+	// Here we simulate the case when the alertmanager receives resolved alerts
+	// right after the aggrGroup is deleted.
+	t.Run("repeated alerts after aggrGroup deleted", func(t *testing.T) {
```

Comment:
Hi! 👋 Josh has asked if I can help take a look at this issue. I'm looking at this test here but I'm not sure I understand what is being tested and how this is showing one of the reported race conditions (I understand there are at least two?).

Here you are adding an expired alert to the store and waiting for a flush. Your test expects no flush to occur, so when it does occur the test fails. What I don't understand is how this relates to the race condition? It is expected that when Alertmanager receives a resolved alert it still creates an aggregation group and flushes it.

Reply:
>the race condition could result in normal alerts being deleted if the same alert becomes active again and is received at the same time as the deletion of the alert.

OK, I think I understand the issue!

Goroutine #1: You have a Resolved alert that is about to be deleted because the if statement is true:

```
if a.Resolved() && got.UpdatedAt == a.UpdatedAt {
	...
}
```

Goroutine #2: Between execution of the condition and execution of the code in the if statement, a firing alert is put into the store via `func (ag *aggrGroup) insert(alert *types.Alert)`, changing the alert back from resolved to firing.

Goroutine #1: The code in the if statement resumes execution and deletes the alert from the store. However, the alert has changed since the condition was execution, and `got` no longer refers to the alert in the store. The firing alert is deleted.

```
if err := ag.alerts.Delete(fp); err != nil {
	...
}
```

> We reject storing resolved alerts if the corresponding firing alert is not found, it may have been lost due to network conditions.

I don't think this fix works though. It has a bug where the Alertmanager does not accept resolved alerts following a crash or restart. Alertmanager does not write its state to disk, so there is no firing alert when the resolved alert is received.

Reply:
>the race condition could result in normal alerts being deleted if the same alert becomes active again and is received at the same time as the deletion of the alert.

OK, I think I understand the issue!

Goroutine #1: You have a Resolved alert that is about to be deleted because the if statement is true:

```
if a.Resolved() && got.UpdatedAt == a.UpdatedAt {
	...
}
```

Goroutine #2: Between execution of the condition and execution of the code in the if statement, a firing alert is put into the store via `func (ag *aggrGroup) insert(alert *types.Alert)`, changing the alert back from resolved to firing.

Goroutine #1: The code in the if statement resumes execution and deletes the alert from the store. However, the alert has changed since the condition was executed, and `got` no longer refers to the alert in the store. The firing alert is deleted.

```
if err := ag.alerts.Delete(fp); err != nil {
	...
}
```

> We reject storing resolved alerts if the corresponding firing alert is not found, it may have been lost due to network conditions.

I don't think this fix works though. It has a bug where the Alertmanager does not accept resolved alerts following a crash or restart. Alertmanager does not write its state to disk, so there is no firing alert when the resolved alert is received.

Reply:
Then, I would prefer to avoid receiving an orphan resolved alert, and I think that might be why we have an option called `resolved_timeout`.

Reply:
What do you mean by orphaned alert?

Reply:
In this case, AlertManager could deliver a resolved alert without a preceding firing alert, so I refer to it as an 'orphan resolved alert'.

Reply:
I think the resolved alert should not be too late, so we can manage the restart scenario efficiently by comparing the process start time with the creation time of the aggregation group.

Reply:
I think that will be difficult to avoid for a number of reasons. Let's still fix the race condition though.

Reply:
OK! I am able to reproduce the issue with a unit test.

The test creates and starts a dispatcher, inserts a firing alert into the store, and waits for the alert to be flushed. It then inserts a resolved alert and waits for it to be flushed too.

Last it inserts the same firing alert that it inserted at the start of the test and waits for that to be flushed. However, the flush never happens because the firing alert is deleted.

```
=== RUN   TestDispatcherRaceFiringAlertDeleted
level=debug component=dispatcher msg="Received alert" alert=[3fff2c2][active]
level=debug component=dispatcher aggrGroup={}:{} msg=flushing alerts=[[3fff2c2][active]]
level=debug component=dispatcher msg="Received alert" alert=[3fff2c2][resolved]
level=debug component=dispatcher aggrGroup={}:{} msg=flushing alerts=[[3fff2c2][resolved]]
level=debug component=dispatcher msg="Received alert" alert=[3fff2c2][active]
level=debug component=dispatcher aggrGroup={}:{} msg="Deleted alert" alert=[3fff2c2][active]
    dispatch_test.go:693:
--- FAIL: TestDispatcherRaceFiringAlertDeleted (3.00s)
FAIL
FAIL	github.com/prometheus/alertmanager/dispatch	3.260s
FAIL
```

To make the race condition easier to observe I added a 1ms sleep to the dispatcher between the call to `Get(fp)` and `Delete(fp)`:

```go
diff --git a/dispatch/dispatch.go b/dispatch/dispatch.go
index 104471e3..c2669496 100644
--- a/dispatch/dispatch.go
+++ b/dispatch/dispatch.go
@@ -526,10 +526,16 @@ func (ag *aggrGroup) flush(notify func(...*types.Alert) bool) {
                                level.Error(ag.logger).Log("msg", "failed to get alert", "err", err, "alert", a.String())
                                continue
                        }
+                       // Insert a 1ms sleep. This allows the scheduler to resume the test
+                       // between ag.alerts.Get(fp) and ag.alerts.Delete(fp) during which
+                       // it should replace the resolved alert with a firing alert.
+                       <-time.After(time.Millisecond)
                        if a.Resolved() && got.UpdatedAt == a.UpdatedAt {
-                               if err := ag.alerts.Delete(fp); err != nil {
+                               deleted, err := ag.alerts.Delete(fp)
+                               if err != nil {
                                        level.Error(ag.logger).Log("msg", "error on delete alert", "err", err, "alert", a.String())
                                }
+                               level.Info(ag.logger).Log("msg", "Deleted alert", "alert", deleted)
                        }
                }
        }
```

I also updated the store to return the deleted alert:

```go
diff --git a/store/store.go b/store/store.go
index 9b30542f..2d5e4666 100644
--- a/store/store.go
+++ b/store/store.go
@@ -115,12 +115,12 @@ func (a *Alerts) Set(alert *types.Alert) error {
 }

 // Delete removes the Alert with the matching fingerprint from the store.
-func (a *Alerts) Delete(fp model.Fingerprint) error {
+func (a *Alerts) Delete(fp model.Fingerprint) (*types.Alert, error) {
        a.Lock()
        defer a.Unlock()
-
+       tmp := a.c[fp]
        delete(a.c, fp)
-       return nil
+       return tmp, nil
 }
```

We can see that the the firing alert is deleted:

```
level=debug component=dispatcher aggrGroup={}:{} msg=flushing alerts=[[3fff2c2][resolved]]
level=debug component=dispatcher msg="Received alert" alert=[3fff2c2][active]
level=debug component=dispatcher aggrGroup={}:{} msg="Deleted alert" alert=[3fff2c2][active]
```

However, what should happen is no alert is deleted as the alert in the store was updated during the flush.